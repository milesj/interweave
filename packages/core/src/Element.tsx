import React from 'react';

export interface ElementProps {
	[prop: string]: unknown;
	className?: string;
	children?: React.ReactNode;
	selfClose?: boolean;
	tagName: string;
}

export function Element({
	className,
	children = null,
	selfClose = false,
	tagName,
	...props
}: ElementProps) {
	const Tag = tagName as 'span';

	return selfClose ? (
		<Tag className={className} {...props} />
	) : (
		<Tag className={className} {...props}>
			{children}
		</Tag>
	);
}
