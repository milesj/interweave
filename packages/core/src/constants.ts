/* eslint-disable no-bitwise, no-magic-numbers, sort-keys */

import { ConfigMap, FilterMap, NodeConfig } from './types';

// https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories
export const TYPE_FLOW = 1;
export const TYPE_SECTION = 1 << 1;
export const TYPE_HEADING = 1 << 2;
export const TYPE_PHRASING = 1 << 3;
export const TYPE_EMBEDDED = 1 << 4;
export const TYPE_INTERACTIVE = 1 << 5;
export const TYPE_PALPABLE = 1 << 6;

// https://developer.mozilla.org/en-US/docs/Web/HTML/Element
const tagConfigs: { [tagName: string]: Partial<NodeConfig> } = {
  a: {
    content: TYPE_FLOW | TYPE_PHRASING,
    self: false,
    type: TYPE_FLOW | TYPE_PHRASING | TYPE_INTERACTIVE | TYPE_PALPABLE,
  },
  address: {
    invalid: [
      'h1',
      'h2',
      'h3',
      'h4',
      'h5',
      'h6',
      'address',
      'article',
      'aside',
      'section',
      'div',
      'header',
      'footer',
    ],
    self: false,
  },
  audio: {
    children: ['track', 'source'],
  },
  br: {
    type: TYPE_FLOW | TYPE_PHRASING,
    void: true,
  },
  body: {
    content:
      TYPE_FLOW |
      TYPE_SECTION |
      TYPE_HEADING |
      TYPE_PHRASING |
      TYPE_EMBEDDED |
      TYPE_INTERACTIVE |
      TYPE_PALPABLE,
  },
  button: {
    content: TYPE_PHRASING,
    type: TYPE_FLOW | TYPE_PHRASING | TYPE_INTERACTIVE | TYPE_PALPABLE,
  },
  caption: {
    content: TYPE_FLOW,
    parent: ['table'],
  },
  col: {
    parent: ['colgroup'],
    void: true,
  },
  colgroup: {
    children: ['col'],
    parent: ['table'],
  },
  details: {
    children: ['summary'],
    type: TYPE_FLOW | TYPE_INTERACTIVE | TYPE_PALPABLE,
  },
  dd: {
    content: TYPE_FLOW,
    parent: ['dl'],
  },
  dl: {
    children: ['dt', 'dd'],
    type: TYPE_FLOW,
  },
  dt: {
    content: TYPE_FLOW,
    invalid: ['footer', 'header'],
    parent: ['dl'],
  },
  figcaption: {
    content: TYPE_FLOW,
    parent: ['figure'],
  },
  footer: {
    invalid: ['footer', 'header'],
  },
  header: {
    invalid: ['footer', 'header'],
  },
  hr: {
    type: TYPE_FLOW,
    void: true,
  },
  img: {
    void: true,
  },
  li: {
    content: TYPE_FLOW,
    parent: ['ul', 'ol', 'menu'],
  },
  main: {
    self: false,
  },
  ol: {
    children: ['li'],
    type: TYPE_FLOW,
  },
  picture: {
    children: ['source', 'img'],
    type: TYPE_FLOW | TYPE_PHRASING | TYPE_EMBEDDED,
  },
  rb: {
    parent: ['ruby', 'rtc'],
  },
  rp: {
    parent: ['ruby', 'rtc'],
  },
  rt: {
    content: TYPE_PHRASING,
    parent: ['ruby', 'rtc'],
  },
  rtc: {
    content: TYPE_PHRASING,
    parent: ['ruby'],
  },
  ruby: {
    children: ['rb', 'rp', 'rt', 'rtc'],
  },
  source: {
    parent: ['audio', 'video', 'picture'],
    void: true,
  },
  summary: {
    content: TYPE_PHRASING,
    parent: ['details'],
  },
  table: {
    children: ['caption', 'colgroup', 'thead', 'tbody', 'tfoot', 'tr'],
    type: TYPE_FLOW,
  },
  tbody: {
    parent: ['table'],
    children: ['tr'],
  },
  td: {
    content: TYPE_FLOW,
    parent: ['tr'],
  },
  tfoot: {
    parent: ['table'],
    children: ['tr'],
  },
  th: {
    content: TYPE_FLOW,
    parent: ['tr'],
  },
  thead: {
    parent: ['table'],
    children: ['tr'],
  },
  tr: {
    parent: ['table', 'tbody', 'thead', 'tfoot'],
    children: ['th', 'td'],
  },
  track: {
    parent: ['audio', 'video'],
    void: true,
  },
  ul: {
    children: ['li'],
    type: TYPE_FLOW,
  },
  video: {
    children: ['track', 'source'],
  },
  wbr: {
    type: TYPE_FLOW | TYPE_PHRASING,
    void: true,
  },
};

function createConfigBuilder(config: Partial<NodeConfig>): (tagName: string) => void {
  return (tagName: string) => {
    tagConfigs[tagName] = {
      ...config,
      ...tagConfigs[tagName],
    };
  };
}

['address', 'main', 'div', 'figure', 'p', 'pre'].forEach(
  createConfigBuilder({
    content: TYPE_FLOW,
    type: TYPE_FLOW | TYPE_PALPABLE,
  }),
);

[
  'abbr',
  'b',
  'bdi',
  'bdo',
  'cite',
  'code',
  'data',
  'dfn',
  'em',
  'i',
  'kbd',
  'mark',
  'q',
  'ruby',
  'samp',
  'strong',
  'sub',
  'sup',
  'time',
  'u',
  'var',
].forEach(
  createConfigBuilder({
    content: TYPE_PHRASING,
    type: TYPE_FLOW | TYPE_PHRASING | TYPE_PALPABLE,
  }),
);

['p', 'pre'].forEach(
  createConfigBuilder({
    content: TYPE_PHRASING,
    type: TYPE_FLOW | TYPE_PALPABLE,
  }),
);

['s', 'small', 'span', 'del', 'ins'].forEach(
  createConfigBuilder({
    content: TYPE_PHRASING,
    type: TYPE_FLOW | TYPE_PHRASING,
  }),
);

['article', 'aside', 'footer', 'header', 'nav', 'section', 'blockquote'].forEach(
  createConfigBuilder({
    content: TYPE_FLOW,
    type: TYPE_FLOW | TYPE_SECTION | TYPE_PALPABLE,
  }),
);

['h1', 'h2', 'h3', 'h4', 'h5', 'h6'].forEach(
  createConfigBuilder({
    content: TYPE_PHRASING,
    type: TYPE_FLOW | TYPE_HEADING | TYPE_PALPABLE,
  }),
);

['audio', 'canvas', 'iframe', 'img', 'video'].forEach(
  createConfigBuilder({
    type: TYPE_FLOW | TYPE_PHRASING | TYPE_EMBEDDED | TYPE_PALPABLE,
  }),
);

// Disable this map from being modified
export const TAGS: ConfigMap = Object.freeze(tagConfigs);

// Tags that should never be allowed, even if the allow list is disabled
export const BANNED_TAG_LIST = [
  'applet',
  'base',
  'body',
  'command',
  'embed',
  'frame',
  'frameset',
  'head',
  'html',
  'link',
  'meta',
  'noscript',
  'object',
  'script',
  'style',
  'title',
];

export const ALLOWED_TAG_LIST = Object.keys(TAGS).filter(
  (tag) => tag !== 'canvas' && tag !== 'iframe',
);

// Filters apply to HTML attributes
export const FILTER_ALLOW = 1;
export const FILTER_DENY = 2;
export const FILTER_CAST_NUMBER = 3;
export const FILTER_CAST_BOOL = 4;
export const FILTER_NO_CAST = 5;

// Attributes not listed here will be denied
// https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes
export const ATTRIBUTES: FilterMap = Object.freeze({
  alt: FILTER_ALLOW,
  cite: FILTER_ALLOW,
  class: FILTER_ALLOW,
  colspan: FILTER_CAST_NUMBER,
  controls: FILTER_CAST_BOOL,
  datetime: FILTER_ALLOW,
  default: FILTER_CAST_BOOL,
  disabled: FILTER_CAST_BOOL,
  dir: FILTER_ALLOW,
  height: FILTER_ALLOW,
  href: FILTER_ALLOW,
  id: FILTER_ALLOW,
  kind: FILTER_ALLOW,
  label: FILTER_ALLOW,
  lang: FILTER_ALLOW,
  loading: FILTER_ALLOW,
  loop: FILTER_CAST_BOOL,
  media: FILTER_ALLOW,
  muted: FILTER_CAST_BOOL,
  poster: FILTER_ALLOW,
  role: FILTER_ALLOW,
  rowspan: FILTER_CAST_NUMBER,
  scope: FILTER_ALLOW,
  sizes: FILTER_ALLOW,
  span: FILTER_CAST_NUMBER,
  start: FILTER_CAST_NUMBER,
  style: FILTER_NO_CAST,
  src: FILTER_ALLOW,
  srclang: FILTER_ALLOW,
  srcset: FILTER_ALLOW,
  target: FILTER_ALLOW,
  title: FILTER_ALLOW,
  type: FILTER_ALLOW,
  width: FILTER_ALLOW,
});

// Attributes to camel case for React props
export const ATTRIBUTES_TO_PROPS: { [key: string]: string } = Object.freeze({
  class: 'className',
  colspan: 'colSpan',
  datetime: 'dateTime',
  rowspan: 'rowSpan',
  srclang: 'srcLang',
  srcset: 'srcSet',
});
