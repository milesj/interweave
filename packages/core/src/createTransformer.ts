import { CommonInternals, Node, PassthroughProps, TagName, WildTagName } from './types';

export type InferElement<K> = K extends '*'
	? HTMLElement
	: K extends keyof HTMLElementTagNameMap
	? HTMLElementTagNameMap[K]
	: HTMLElement;

export type TransformerFactory<Element, Props extends object> = (
	element: Element,
	props: Props,
	children: Node[],
) => Element | React.ReactElement | null | undefined | void;

export interface TransformerOptions<Props extends object, Config extends object>
	extends CommonInternals<Props, Config> {
	tagName?: TagName;
}

export interface Transformer<Element, Props extends object, Config extends object>
	extends CommonInternals<Props, Config> {
	extend: (
		config?: Partial<Config>,
		factory?: TransformerFactory<Element, Props>,
	) => Transformer<Element, Props, Config>;
	factory: TransformerFactory<Element, Props>;
	tagName: WildTagName;
}

export function createTransformer<
	K extends WildTagName,
	Props extends object = PassthroughProps,
	Config extends object = {},
>(
	tagName: K,
	factory: TransformerFactory<InferElement<K>, Props>,
	options: TransformerOptions<Props, Config> = {},
): Transformer<InferElement<K>, Props, Config> {
	return {
		config: options.config,
		extend(customConfig, customFactory) {
			return createTransformer<K, Props, Config>(tagName, customFactory ?? factory, {
				...options,
				config: {
					...options.config!,
					...customConfig,
				},
			});
		},
		factory,
		onAfterParse: options.onAfterParse,
		onBeforeParse: options.onBeforeParse,
		tagName: options.tagName ?? tagName,
	};
}
