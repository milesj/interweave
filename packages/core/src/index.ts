/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './constants';
export * from './createMatcher';
export * from './createTransformer';
export * from './Element';
export * from './Interweave';
export * from './Markup';
export * from './Parser';
export * from './types';
