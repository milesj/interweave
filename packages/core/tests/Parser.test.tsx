import React from 'react';
import {
	ATTRIBUTES,
	ATTRIBUTES_TO_PROPS,
	FILTER_ALLOW,
	FILTER_CAST_BOOL,
	FILTER_CAST_NUMBER,
} from '../src/constants';
import { Element } from '../src/Element';
import { Parser } from '../src/Parser';
import {
	codeBarMatcher,
	codeBazMatcher,
	codeFooMatcher,
	createExpectedToken,
	linkTransformer,
	MOCK_MARKUP,
	parentConfig,
	TOKEN_LOCATIONS,
} from '../src/test';

function createChild(tag: string, text: number | string): HTMLElement {
	const child = document.createElement(tag);
	child.append(document.createTextNode(String(text)));

	return child;
}

describe('Parser', () => {
	let instance: Parser;
	let element: HTMLElement;

	beforeEach(() => {
		instance = new Parser(
			'',
			{},
			[codeFooMatcher, codeBarMatcher, codeBazMatcher],
			[linkTransformer],
		);
	});

	it('parses when passed an empty value', () => {
		[null, false, undefined, '', 0].forEach((value) => {
			// @ts-expect-error Invalid type
			expect(new Parser(value).parse()).toBeNull();
		});
	});

	it('errors for a non-string value', () => {
		[123, 456.78, true, [], {}].forEach((value) => {
			// @ts-expect-error Invalid type
			expect(() => new Parser(value)).toThrow('Interweave parser requires a valid string.');
		});
	});

	describe('applyMatchers()', () => {
		function createElement(value: string, key: number) {
			return (
				<Element key={key} customprop={value} tagName="span">
					{value.toUpperCase()}
				</Element>
			);
		}

		function createMultiElement(value: string, key: number) {
			switch (key) {
				case 1:
					return createElement('bar', 1);
				case 2:
					return createElement('baz', 2);
				default:
					return createElement('foo', 0);
			}
		}

		describe('handles matchers correctly', () => {
			TOKEN_LOCATIONS.forEach((location, i) => {
				it(`for: ${location}`, () => {
					instance.keyIndex = -1; // Reset for easier testing

					const tokenString = location.replace(/{token}/g, '[foo]');
					const actual = instance.applyMatchers(tokenString, parentConfig);

					if (i === 0) {
						expect(actual).toBe(createExpectedToken('foo', createElement, 0));
					} else {
						expect(actual).toEqual(createExpectedToken('foo', createElement, i));
					}
				});
			});
		});

		describe('handles multiple matchers correctly', () => {
			TOKEN_LOCATIONS.forEach((location, i) => {
				it(`for: ${location}`, () => {
					instance.keyIndex = -1; // Reset for easier testing

					const tokenString = location
						.replace('{token}', '[foo]')
						.replace('{token}', '[bar]')
						.replace('{token}', '[baz]');
					const actual = instance.applyMatchers(tokenString, parentConfig);

					if (i === 0) {
						expect(actual).toBe(createExpectedToken('', createMultiElement, 0));
					} else {
						expect(actual).toEqual(createExpectedToken('', createMultiElement, i));
					}
				});
			});
		});

		describe('handles no matchers correctly', () => {
			TOKEN_LOCATIONS.forEach((location, i) => {
				it(`for: ${location}`, () => {
					const tokenString = location.replace(/{token}/g, '[qux]');
					const actual = instance.applyMatchers(tokenString, parentConfig);

					expect(actual).toBe(createExpectedToken('[qux]', (value) => value, i, true));
				});
			});
		});
	});

	describe('canRenderChild()', () => {
		it('doesnt render if missing parent tag', () => {
			expect(
				instance.canRenderChild(
					{
						...parentConfig,
						// @ts-expect-error Invalid
						tagName: '',
					},
					{ ...parentConfig },
				),
			).toBe(false);
		});

		it('doesnt render if missing child tag', () => {
			expect(
				instance.canRenderChild(
					{ ...parentConfig },
					{
						...parentConfig,
						// @ts-expect-error Invalid
						tagName: '',
					},
				),
			).toBe(false);
		});
	});

	describe('createContainer()', () => {
		it('injects the markup into the body', () => {
			const body = instance.createContainer(
				'<div>Foo<section>Bar</section><aside>Baz</aside></div>',
			);

			expect(body?.outerHTML).toBe(
				'<body><div>Foo<section>Bar</section><aside>Baz</aside></div></body>',
			);
		});

		it('errors out for `!DOCTYPE`', () => {
			expect(() => {
				instance.createContainer('<!DOCTYPE><html><body>Foo</body></html>');
			}).toThrow('HTML documents as Interweave content are not supported.');
		});

		it('errors out for `html`', () => {
			expect(() => {
				instance.createContainer('<html><body>Foo</body></html>');
			}).toThrow('HTML documents as Interweave content are not supported.');
		});

		it('errors out for `head`', () => {
			expect(() => {
				instance.createContainer('<head></head><body>Foo</body>');
			}).toThrow('HTML documents as Interweave content are not supported.');
		});

		it('errors out for `body`', () => {
			expect(() => {
				instance.createContainer('<body>Foo</body>');
			}).toThrow('HTML documents as Interweave content are not supported.');
		});

		it('doesnt errors out for `html` in brackets', () => {
			expect(() => {
				instance.createContainer('[html]Body[/html]');
			}).not.toThrow();
		});
	});

	describe('convertLineBreaks()', () => {
		it('doesnt convert when HTML closing tags exist', () => {
			expect(instance.convertLineBreaks('<div>It\nwont\r\nconvert.</div>')).toBe(
				'<div>It\nwont\r\nconvert.</div>',
			);
		});

		it('doesnt convert when HTML void tags exist', () => {
			expect(instance.convertLineBreaks('It\n<br/>wont\r\nconvert.')).toBe(
				'It\n<br/>wont\r\nconvert.',
			);
		});

		it('doesnt convert when HTML void tags with spaces exist', () => {
			expect(instance.convertLineBreaks('It\n<br  />wont\r\nconvert.')).toBe(
				'It\n<br  />wont\r\nconvert.',
			);
		});

		it('doesnt convert if `noHtml` is true', () => {
			instance.props.noHtml = true;

			expect(instance.convertLineBreaks('It\nwont\r\nconvert.')).toBe('It\nwont\r\nconvert.');
		});

		it('doesnt convert if `disableLineBreaks` is true', () => {
			instance.props.disableLineBreaks = true;

			expect(instance.convertLineBreaks('It\nwont\r\nconvert.')).toBe('It\nwont\r\nconvert.');
		});

		it('replaces carriage returns', () => {
			expect(instance.convertLineBreaks('Foo\r\nBar')).toBe('Foo<br/>Bar');
		});

		it('replaces super long multilines', () => {
			expect(instance.convertLineBreaks('Foo\n\n\n\n\n\n\nBar')).toBe('Foo<br/><br/><br/>Bar');
		});
	});

	describe('extractAttributes()', () => {
		beforeEach(() => {
			element = document.createElement('div');
		});

		it('returns null for invalid node', () => {
			// @ts-expect-error Invalid type
			expect(instance.extractAttributes(document.createComment('Comment'))).toBeNull();
		});

		Object.keys(ATTRIBUTES).forEach((name) => {
			const attrName = ATTRIBUTES_TO_PROPS[name] || name;

			switch (ATTRIBUTES[name]) {
				case FILTER_ALLOW:
					it(`allows the "${name}" attribute and casts to a string`, () => {
						element.setAttribute(name, 'Foo');

						expect(instance.extractAttributes(element)).toEqual({
							[attrName]: 'Foo',
						});
					});
					break;

				case FILTER_CAST_BOOL:
					it(`allows the "${name}" attribute and casts to a boolean`, () => {
						element.setAttribute(name, 'true');

						expect(instance.extractAttributes(element)).toEqual({
							[attrName]: true,
						});

						// Falsy attributes are still enabled according to HTML spec
						element.setAttribute(name, 'false');

						expect(instance.extractAttributes(element)).toEqual({
							[attrName]: true,
						});
					});

					it(`allows the "${name}" attribute and casts to a boolean when setting value equal to name`, () => {
						element.setAttribute(name, name);

						expect(instance.extractAttributes(element)).toEqual({
							[attrName]: true,
						});

						element.setAttribute(name, '');

						expect(instance.extractAttributes(element)).toEqual({
							[attrName]: true,
						});
					});
					break;

				case FILTER_CAST_NUMBER:
					it(`allows the "${name}" attribute and casts to a number`, () => {
						element.setAttribute(name, '100');

						expect(instance.extractAttributes(element)).toEqual({
							[attrName]: 100,
						});
					});
					break;

				default:
					break;
			}
		});

		it('allows aria attributes', () => {
			element.setAttribute('aria-live', 'off');

			expect(instance.extractAttributes(element)).toEqual({
				'aria-live': 'off',
			});
		});

		it('allows data attributes', () => {
			element.dataset.foo = 'bar';

			expect(instance.extractAttributes(element)).toEqual({
				'data-foo': 'bar',
			});
		});

		it('denies attributes that are not allowed', () => {
			element.setAttribute('readonly', 'true');

			expect(instance.extractAttributes(element)).toBeNull();
		});

		it('denies attributes that start with on', () => {
			element.setAttribute('onload', 'hackServers();');
			element.setAttribute('onclick', 'doSomething();');
			// eslint-disable-next-line unicorn/prefer-add-event-listener
			element.onmouseenter = function noop() {};

			expect(instance.extractAttributes(element)).toBeNull();
		});

		// https://www.owasp.org/index.php/XSS_Filter_Evasion_Cheat_Sheet
		it('denies sources that have injections', () => {
			/* eslint-disable no-script-url */
			element = document.createElement('a');

			element.setAttribute('href', 'javascript:alert();');
			expect(instance.extractAttributes(element)).toBeNull();

			element.setAttribute('href', 'jav\ras\ncript\t:alert();');
			expect(instance.extractAttributes(element)).toBeNull();

			element.setAttribute('href', 'jav&#x09;ascript:alert();');
			expect(instance.extractAttributes(element)).toBeNull();

			element.setAttribute('href', 'jav&#x0A;ascript:alert();');
			expect(instance.extractAttributes(element)).toBeNull();

			element.setAttribute('href', 'jav&#x0D;ascript:alert();');
			expect(instance.extractAttributes(element)).toBeNull();

			element.setAttribute('href', 'java\0script:alert();');
			expect(instance.extractAttributes(element)).toBeNull();

			element = document.createElement('img');

			element.setAttribute('src', 'javaScript:void(0);');
			expect(instance.extractAttributes(element)).toBeNull();

			element.setAttribute('src', ' &#14;  javascript:void(0);');
			expect(instance.extractAttributes(element)).toBeNull();

			element.setAttribute('src', 'xss:confirm();');
			expect(instance.extractAttributes(element)).toBeNull();
			/* eslint-enable no-script-url */
		});

		it('camel cases specific attribute names to React attribute names', () => {
			element.setAttribute('datetime', '2016-01-01');
			element.setAttribute('colspan', '3');
			element.setAttribute('rowspan', '6');
			element.setAttribute('class', 'foo-bar');
			element.setAttribute('alt', 'Foo');
			element.setAttribute('disabled', 'disabled');

			expect(instance.extractAttributes(element)).toEqual({
				dateTime: '2016-01-01',
				colSpan: 3,
				rowSpan: 6,
				className: 'foo-bar',
				alt: 'Foo',
				disabled: true,
			});
		});

		it('converts `style` to an object', () => {
			element.setAttribute('style', 'background-color: red; color: black; display: inline-block;');

			expect(instance.extractAttributes(element)).toEqual({
				style: {
					backgroundColor: 'red',
					color: 'black',
					display: 'inline-block',
				},
			});
		});
	});

	describe('isTagAllowed()', () => {
		it('always returns false for banned tags', () => {
			instance.allowed.add('script');
			instance.props.allowElements = true;

			expect(instance.isTagAllowed('script')).toBe(false);
		});

		it('always returns false for blocked tags', () => {
			instance.blocked.add('b');
			instance.allowed.add('b');
			instance.props.allowElements = true;

			expect(instance.isTagAllowed('b')).toBe(false);
		});

		it('returns true if `allowElements` but not in allow list', () => {
			instance.allowed.delete('b');
			instance.props.allowElements = true;

			expect(instance.isTagAllowed('b')).toBe(true);
		});

		it('returns true if in allow list', () => {
			// @ts-expect-error Invalid
			instance.allowed.add('custom');

			// @ts-expect-error Invalid
			expect(instance.isTagAllowed('custom')).toBe(true);
		});

		it('returns false if not in allow list', () => {
			instance.allowed.delete('b');

			expect(instance.isTagAllowed('b')).toBe(false);
		});

		it('returns true if in default list', () => {
			instance.allowed.add('span');

			expect(instance.isTagAllowed('span')).toBe(true);
		});
	});

	describe('parse()', () => {
		it('parses the entire document starting from the body', () => {
			instance = new Parser(MOCK_MARKUP, {});

			expect(instance.parse()).toMatchSnapshot();
		});

		it('doesnt parse HTML when escaped', () => {
			instance = new Parser(MOCK_MARKUP, { escapeHtml: true });

			expect(instance.parse()).toMatchSnapshot();
		});
	});

	describe('parseNode()', () => {
		beforeEach(() => {
			element = document.createElement('div');
		});

		it('returns an empty array when no child nodes present', () => {
			expect(instance.parseNode(element, parentConfig)).toEqual([]);
		});

		it('returns text nodes as strings', () => {
			element.append(document.createTextNode('Foo'));
			element.append(document.createTextNode('Bar'));

			expect(instance.parseNode(element, parentConfig)).toEqual(['FooBar']);
		});

		it('ignores unknown elements', () => {
			element.append(document.createElement('foo'));

			expect(instance.parseNode(element, parentConfig)).toEqual([]);
		});

		it('returns text and element nodes in order', () => {
			element.append(document.createTextNode('Foo'));
			element.append(createChild('div', 'Bar'));
			element.append(document.createTextNode('Baz'));

			expect(instance.parseNode(element, parentConfig)).toEqual([
				'Foo',
				<Element key="0" tagName="div">
					{['Bar']}
				</Element>,
				'Baz',
			]);
		});

		it('combines multiple strings together', () => {
			element.append(document.createTextNode('Foo'));
			element.append(createChild('div', 'Bar'));
			element.append(document.createTextNode('Baz'));
			element.append(document.createTextNode('Qux'));
			element.append(createChild('div', 'Bar'));

			expect(instance.parseNode(element, parentConfig)).toEqual([
				'Foo',
				<Element key="0" tagName="div">
					{['Bar']}
				</Element>,
				'BazQux',
				<Element key="1" tagName="div">
					{['Bar']}
				</Element>,
			]);
		});

		it('ignores comment nodes', () => {
			element.append(document.createComment('Comment'));

			expect(instance.parseNode(element, parentConfig)).toEqual([]);
		});

		// it('ignores document nodes', () => {
		//   element.append(document);

		//   expect(instance.parseNode(element, parentConfig)).toEqual([]);
		// });

		it('ignores document fragment nodes', () => {
			element.append(document.createDocumentFragment());

			expect(instance.parseNode(element, parentConfig)).toEqual([]);
		});

		it('passes through elements if `noHtml` prop is set', () => {
			instance = new Parser('', {
				noHtml: true,
			});

			element.append(document.createTextNode('Foo'));
			element.append(createChild('div', 'Bar'));
			element.append(document.createTextNode('Baz'));
			element.append(createChild('div', 'Qux'));
			element.append(createChild('div', 'Wat'));

			expect(instance.parseNode(element, parentConfig)).toEqual([
				'Foo',
				'Bar',
				'Baz',
				'Qux',
				'Wat',
			]);
		});

		it('passes through elements if `noHtmlExceptInternals` prop is set', () => {
			instance = new Parser('', {
				noHtmlExceptInternals: true,
			});

			element.append(document.createTextNode('Foo'));
			element.append(createChild('div', 'Bar'));
			element.append(document.createTextNode('Baz'));
			element.append(createChild('div', 'Qux'));
			element.append(createChild('div', 'Wat'));

			expect(instance.parseNode(element, parentConfig)).toEqual([
				'Foo',
				'Bar',
				'Baz',
				'Qux',
				'Wat',
			]);
		});

		it('strips matchers HTML if `noHtml` prop is set', () => {
			instance.props.noHtml = true;

			element.append(document.createTextNode('Foo'));
			element.append(createChild('div', 'Bar'));
			element.append(document.createTextNode('[foo]'));

			expect(instance.parseNode(element, parentConfig)).toEqual(['Foo', 'Bar', '[foo]']);
		});

		it('doesnt strip matchers HTML if `noHtmlExceptInternals` prop is set', () => {
			instance.props.noHtmlExceptInternals = true;

			element.append(document.createTextNode('Foo'));
			element.append(createChild('div', 'Bar'));
			element.append(document.createTextNode('[foo]'));

			expect(instance.parseNode(element, parentConfig)).not.toEqual(['Foo', 'Bar', '[foo]']);
		});

		it('only renders allowed children', () => {
			element.append(document.createTextNode('Foo'));
			element.append(createChild('i', 'Bar'));
			element.append(document.createTextNode('Baz'));
			element.append(createChild('b', 'Qux'));
			element.append(createChild('u', 'Wat'));

			expect(
				instance.parseNode(element, {
					...parentConfig,
					children: ['b'],
				}),
			).toEqual([
				'Foo',
				'Bar',
				'Baz',
				<Element key="0" tagName="b">
					{['Qux']}
				</Element>,
				'Wat',
			]);
		});

		it('only renders in allowed parent', () => {
			element.append(document.createTextNode('Foo'));
			element.append(createChild('li', 'Bar'));
			element.append(document.createTextNode('Baz'));
			element.append(createChild('li', 'Qux'));
			element.append(createChild('li', 'Wat'));

			expect(
				instance.parseNode(element, {
					...parentConfig,
				}),
			).toEqual(['Foo', 'Bar', 'Baz', 'Qux', 'Wat']);
		});

		it('doesnt render self children', () => {
			element.append(document.createTextNode('Foo'));
			element.append(createChild('div', 'Bar'));
			element.append(document.createTextNode('Baz'));
			element.append(createChild('span', 'Qux'));
			element.append(createChild('section', 'Wat'));

			expect(
				instance.parseNode(element, {
					...parentConfig,
					self: false,
				}),
			).toEqual([
				'Foo',
				'Bar',
				'Baz',
				<Element key="0" tagName="span">
					{['Qux']}
				</Element>,
				<Element key="1" tagName="section">
					{['Wat']}
				</Element>,
			]);
		});

		it('doesnt render elements where a filter returns null', () => {
			element.append(document.createTextNode('Foo'));
			element.append(createChild('link', 'Bar'));
			element.append(document.createTextNode('Baz'));

			expect(instance.parseNode(element, parentConfig)).toEqual(['Foo', 'Baz']);
		});

		it('does render an `a` tag within inline', () => {
			element = document.createElement('span');
			element.append(document.createTextNode('Foo'));
			element.append(createChild('a', 'Bar'));
			element.append(document.createTextNode('Baz'));

			expect(instance.parseNode(element, instance.getTagConfig('span'))).toEqual([
				'Foo',
				<Element key="0" tagName="a" target="_blank">
					{['Bar']}
				</Element>,
				'Baz',
			]);
		});

		it('does render an `a` tag within block', () => {
			element.append(document.createTextNode('Foo'));
			element.append(createChild('a', 'Bar'));
			element.append(document.createTextNode('Baz'));

			expect(
				instance.parseNode(element, {
					...parentConfig,
				}),
			).toEqual([
				'Foo',
				<Element key="0" tagName="a" target="_blank">
					{['Bar']}
				</Element>,
				'Baz',
			]);
		});

		it('does render inline elements within an `a` tag', () => {
			element = document.createElement('a');
			element.append(document.createTextNode('Foo'));
			element.append(createChild('span', 'Bar'));
			element.append(document.createTextNode('Baz'));

			expect(instance.parseNode(element, instance.getTagConfig('a'))).toEqual([
				'Foo',
				<Element key="0" tagName="span">
					{['Bar']}
				</Element>,
				'Baz',
			]);
		});

		it('does render block elements within an `a` tag', () => {
			element = document.createElement('a');
			element.append(document.createTextNode('Foo'));
			element.append(createChild('div', 'Bar'));
			element.append(document.createTextNode('Baz'));

			expect(instance.parseNode(element, instance.getTagConfig('a'))).toEqual([
				'Foo',
				<Element key="0" tagName="div">
					{['Bar']}
				</Element>,
				'Baz',
			]);
		});

		it('uses parent config for unsupported elements', () => {
			element = document.createElement('span');

			const acronym = document.createElement('acronym');
			acronym.append(createChild('a', 'Link'));

			element.append(acronym);

			expect(instance.parseNode(element, instance.getTagConfig('span'))).toMatchSnapshot();
		});
	});
});
