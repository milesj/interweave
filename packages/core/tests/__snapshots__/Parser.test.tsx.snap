// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser parse() doesnt parse HTML when escaped 1`] = `
Array [
  "<main role=\\"main\\">",
  <Element
    selfClose={true}
    tagName="br"
  />,
  "  Main content",
  <Element
    selfClose={true}
    tagName="br"
  />,
  "  <div>",
  <Element
    selfClose={true}
    tagName="br"
  />,
  "    <a href=\\"#\\">Link</a>",
  <Element
    selfClose={true}
    tagName="br"
  />,
  "    <span class=\\"foo\\">String</span>",
  <Element
    selfClose={true}
    tagName="br"
  />,
  "  </div>",
  <Element
    selfClose={true}
    tagName="br"
  />,
  "</main>",
  <Element
    selfClose={true}
    tagName="br"
  />,
  "<aside id=\\"sidebar\\">",
  <Element
    selfClose={true}
    tagName="br"
  />,
  "  Sidebar content",
  <Element
    selfClose={true}
    tagName="br"
  />,
  "</aside>",
]
`;

exports[`Parser parse() parses the entire document starting from the body 1`] = `
Array [
  <Element
    role="main"
    tagName="main"
  >
    
  Main content
  
    <Element
      tagName="div"
    >
      
    
      <Element
        href="#"
        tagName="a"
      >
        Link
      </Element>
      
    
      <Element
        className="foo"
        tagName="span"
      >
        String
      </Element>
      
  
    </Element>
    

  </Element>,
  "
",
  <Element
    id="sidebar"
    tagName="aside"
  >
    
  Sidebar content

  </Element>,
]
`;

exports[`Parser parseNode() uses parent config for unsupported elements 1`] = `
Array [
  <Element
    tagName="a"
    target="_blank"
  >
    Link
  </Element>,
]
`;
